Initial implemention of JSON migrator for the library files

bcdonovan opened the pull request:
This pull request is not intended to be merge in at this time. The PR is to provide for feedback.

@blakejohnson This is the initial implementation of a JSON migration scheme for the Library files. I have included an example verison_1_to_2 method for the channels file. I don't think it's exactly what you had in mind but should provide an example of what I intended.

blakejohnson commented on LibraryMigrator.py:
```diff
@@ -0,0 +1,142 @@
+
+	def version_1_to_2(self):
+
+		if not self.is_class(self.jsonDict,'ChannelLibrary') or 'channelDict' not in self.jsonDict:
```
It might be nice for this kind of boilerplate stuff to happen automagically. Perhaps it should go in `__init__()`?

blakejohnson commented on LibraryMigrator.py:
```diff
@@ -0,0 +1,142 @@
+				continue
+			frequency = self.jsonDict['channelDict'][pc]['SSBFreq']
+			del self.jsonDict['channelDict'][pc]['SSBFreq']
+			self.jsonDict['channelDict'][lc]['SSBFreq'] = frequency
```
This is pretty close to what I had in mind. I was just going to name this field `frequency` rather than `SSBFreq`.

bcdonovan commented:
Incorporated @blakejohnson's input.
Added migration for labbrick64 -> labbrick
Implemented change in frequency location in the respective channel classes and GUI.
Removed labbrick64 class from MicrowaveSources.

blakejohnson commented:
This looks good, but I'll need to prepare another pull request that builds on this one to incorporate necessary changes in the compiler. So, we need to wait to merge for a little bit.

bcdonovan commented:
Sounds good. I'll set this aside for now then.

blakejohnson commented:
One thing remaining to do here is decide how this should be exposed to end users. Do they call it directly? Do we try to automatically migrate outdated JSON files?

bcdonovan commented:
I think you automatically migrate with a notification. 

The simple way to do this is to run the migrator every time in both startup.py and ExpSettingsGUI.py. 

When run from startup the migrator can print the console. When run from ExpSettingsGUI a pop-up shown. 

Alternatively the individual libraries could test the version of the files on load and then call the migrator. This would couple the libraries to the migrator at run time in addition to design time. The advantage is the migration would be attempted whenever the libraries are loaded.

rblazarus commented:
Would it also be appropriate to issue a deprecation warning when using the old format?

blakejohnson commented:
@rblazarus the way the parameter libraries work, you *cannot* use an old format. It will fail to load. Hence this migrator. So, the best we can do is to warn the user that we have updated their libraries to a new format.

blakejohnson commented:
@bcdonovan you seem to have merged a rebase into this branch rather than replacing the branch itself. Mind if I try to fix it?

blakejohnson commented:
Ah, that's better. I think the same thing happen on your test/GUI branch.

bcdonovan commented:
@blakejohnson I've fixed both of my branches so they rebase off develop post merge #53

blakejohnson commented:
Rebased again so that I can test against the latest changes to the unit tests.
